internal/server/data/destination.go

```diff
@@ -221,3 +222,103 @@ func CountDestinationsByConnectedVersion(tx ReadTxn) ([]DestinationsCount, error
 func CountAllDestinations(tx ReadTxn) (int64, error) {
 	return countRows(tx, destinationsTable{})
 }
+
+type DestinationAccess struct {
+	UserID           uid.ID
+	UserSSHLoginName string
+	Privilege        string
+	Resource         string
+}
+
+type ListDestinationAccessResult struct {
+	// Items is the list of items that match the query, excluding soft-deleted rows.
+	Items []DestinationAccess
+
+	// MaxUpdateIndex is the maximum update_index from all the grants and groups
+	// that match the query. This MUST include soft-deleted rows as well.
+	// The value will be 1 if no records match the query, so that the caller can
+	// block until a record exists.
+	MaxUpdateIndex int64
+}
+
+func ListDestinationAccess(tx ReadTxn, destination string) (*ListDestinationAccessResult, error) {
+	// IMPORTANT: changes to this query likely also need to be applied to
+	// destinationAccessMaxUpdateIndex.
+	query := querybuilder.New("SELECT")
```

Comment:
this is pretty difficult to read, but I don't think there's a better way to do this with our table structure 👍